homeassistant/components/smartthings/smartapp.py

```diff
@@ -0,0 +1,266 @@
+        hass.data[DOMAIN][CONF_INSTANCE_ID]
+    await api.update_app_settings(settings)
+    _LOGGER.debug("Updated App Settings for SmartApp '%s' (%s).",
+                  app.app_name, app.app_id)
+
+    # Set oauth scopes
+    oauth = AppOAuth(app.app_id)
+    oauth.client_name = 'Home Assistant'
+    oauth.scope.extend(APP_OAUTH_SCOPES)
+    await api.update_app_oauth(oauth)
+    _LOGGER.debug("Updated App OAuth for SmartApp '%s' (%s).",
+                  app.app_name, app.app_id)
+    return app
+
+
+async def update_app(hass: HomeAssistantType, app):
+    """Ensure the SmartApp is up-to-date and update if necessary."""
+    template = _get_app_template(hass)
+    template.pop('app_name')  # don't update this
+    update_required = False
+    for key, value in template.items():
+        if getattr(app, key) != value:
+            update_required = True
+            setattr(app, key, value)
+    if update_required:
+        await app.save()
+        _LOGGER.debug("SmartApp '%s' (%s) updated with latest settings.",
+                      app.app_name, app.app_id)
+
+
+def setup_smartapp(hass, app):
+    """
+    Configure an individual SmartApp in hass.
+
+    Register the SmartApp with the SmartAppManager so that hass will service
+    lifecycle events (install, event, etc...).  A unique SmartApp is created
+    for each SmartThings account that is configured in hass.
+    """
+    manager = hass.data[DOMAIN][DATA_MANAGER]
+    smartapp = manager.smartapps.get(app.app_id)
+    if smartapp:
+        # already setup
+        return smartapp
+    smartapp = manager.register(app.app_id, app.webhook_public_key)
+    smartapp.name = app.display_name
+    smartapp.description = app.description
+    smartapp.permissions.extend(APP_OAUTH_SCOPES)
+    return smartapp
+
+
+async def setup_smartapp_endpoint(hass: HomeAssistantType):
+    """
+    Configure the SmartApp webhook in hass.
+
+    SmartApps are an extension point within the SmartThings ecosystem and
+    is used to receive push updates (i.e. device updates) from the cloud.
+    """
+    from pysmartapp import Dispatcher, SmartAppManager
+
+    data = hass.data.get(DOMAIN)
+    if data:
+        # already setup
+        return
+
+    # Get/create config to store a unique id for this hass instance.
+    path = hass.config.path(SMARTTHINGS_CONFIG_FILE)
+    if not await hass.async_add_job(os.path.isfile, path):
+        # Create config
+        config = {
+            CONF_INSTANCE_ID: str(uuid4()),
+            CONF_WEBHOOK_ID: webhook.generate_secret()
+        }
+        await hass.async_add_job(save_json, path, config)
+    else:
+        config = await hass.async_add_job(load_json, path)
+
+    # SmartAppManager uses a dispatcher to invoke callbacks when push events
+    # occur. Use hass' implementation instead of the built-in one.
+    dispatcher = Dispatcher(
+        signal_prefix=SIGNAL_SMARTAPP_PREFIX,
+        connect=functools.partial(async_dispatcher_connect, hass),
```

Comment:
connect/send are async?

Reply:
Yes.  The [Dispatcher](https://github.com/andrewsayre/pysmartapp/blob/master/pysmartapp/dispatch.py) implementation is almost the same as HASS' except it doesn't track tasks.